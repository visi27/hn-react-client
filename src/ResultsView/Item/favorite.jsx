import { PropTypes } from 'prop-types';
import React, { Component } from 'react';
import { Icon } from 'react-icons-kit';
import { heart } from 'react-icons-kit/fa/';
import ReactLoading from 'react-loading';

const IconHeart = props => <Icon icon={heart} {...props} />;
const Loading = <ReactLoading type="spin" color="#F4A261" height={18} width={18} />;

class Favorite extends Component {
  constructor(props) {
    super(props);

    const { itemId, isFavorite } = props;

    this.state = {
      isLoading: false,
      isFavorite: isFavorite(itemId),
    };
  }

  addToFavorites() {
    this.setState({ isLoading: true });
  }

  render() {
    const { itemId, onAddFavorite, onRemoveFavorite } = this.props;
    const { isLoading, isFavorite } = this.state;
    const FavIcon = isFavorite ? (
      <IconHeart
        style={{ color: '#F4A261' }}
        onClick={() => {
          this.setState({ isLoading: true });
          onRemoveFavorite(itemId);
          setTimeout(() => this.setState({ isLoading: false, isFavorite: false }), 1000);
          // this.forceUpdate();
        }}
      />
    ) : (
      <IconHeart
        onClick={() => {
          this.setState({ isLoading: true });
          onAddFavorite(itemId);
          setTimeout(() => this.setState({ isLoading: false, isFavorite: true }), 1000);
          // this.forceUpdate();
        }}
      />
    );

    return isLoading ? Loading : FavIcon;
  }
}

Favorite.propTypes = {
  itemId: PropTypes.string.isRequired,
  isFavorite: PropTypes.func,
  onAddFavorite: PropTypes.func,
  onRemoveFavorite: PropTypes.func,
};

Favorite.defaultProps = {
  isFavorite: () => {},
  onAddFavorite: () => {},
  onRemoveFavorite: () => {},
};

export default Favorite;
