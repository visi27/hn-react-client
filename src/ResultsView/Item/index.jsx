import moment from 'moment';
import { PropTypes } from 'prop-types';
import React, { Component } from 'react';
import { IconButton } from 'react-toolbox/lib/button';

class Item extends Component {
  render() {
    const {
      item, isFavorite, onAddFavorite, onRemoveFavorite,
    } = this.props;
    return (
      <div key={item.objectID} className="container-fluid item">
        <div className="list-inline">
          <span className="list-inline-item">
            <a className="h5 text-muted" href={item.url}>
              {item.title}
            </a>
          </span>
          <span className="list-inline-item">
            {isFavorite(item.objectID) ? (
              <IconButton
                icon="favorite"
                accent
                onClick={() => {
                  onRemoveFavorite(item.objectID);
                  this.forceUpdate();
                }}
              />
            ) : (
              <IconButton
                icon="favorite"
                onClick={() => {
                  onAddFavorite(item.objectID);
                  this.forceUpdate();
                }}
              />
            )}
          </span>
        </div>
        <div className="list-inline">
          <span className="list-inline-item">Author: {item.author}</span>
          <span className="list-inline-item">Created: {moment(item.created_at).fromNow()}</span>
          <span className="list-inline-item">Comments: {item.num_comments}</span>
          <span className="list-inline-item">Points: {item.points}</span>
          <span className="list-inline-item">
            <a className="text-muted" href={item.url}>
              ({item.url})
            </a>
          </span>
        </div>
        <hr />
      </div>
    );
  }
}

Item.propTypes = {
  item: PropTypes.shape({
    created_at: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    title_highlighted: PropTypes.string,
    url: PropTypes.string,
    author: PropTypes.string.isRequired,
    points: PropTypes.number.isRequired,
    story_text: PropTypes.string,
    num_comments: PropTypes.number.isRequired,
    objectID: PropTypes.string.isRequired,
  }).isRequired,
  isFavorite: PropTypes.func,
  onAddFavorite: PropTypes.func,
  onRemoveFavorite: PropTypes.func,
};

Item.defaultProps = {
  isFavorite: () => {},
  onAddFavorite: () => {},
  onRemoveFavorite: () => {},
};

export default Item;
