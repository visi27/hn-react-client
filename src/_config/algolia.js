const algolia = {
  DEFAULT_QUERY: '',
  DEFAULT_HPP: '20',

  PATH_BASE: 'https://hn.algolia.com/api/v1/',
  PATH_SEARCH: '/search',
  PARAM_SEARCH: 'query=',
  PARAM_PAGE: 'page=',
  PARAM_HPP: 'hitsPerPage=',
};

export default algolia;
