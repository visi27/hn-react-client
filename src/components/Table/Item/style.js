export const largeColumn = {
  width: '40%',
};
export const midColumn = {
  width: '30%',
};
export const smallColumn = {
  width: '10%',
};
