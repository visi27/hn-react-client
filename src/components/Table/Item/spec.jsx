import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Item from './index';

describe('Item', () => {
  const item = {
    title: '1',
    author: '1',
    num_comments: 1,
    points: 2,
    objectID: 'y',
    created_at: '2018-04-03T09:35:47.000Z',
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Item item={item} />, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<Item item={item} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
