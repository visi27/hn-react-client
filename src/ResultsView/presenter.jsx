import { PropTypes } from 'prop-types';
import React, { Component } from 'react';
import ReactLoading from 'react-loading';
import Paginator from '../Paginator/index';
import algolia from '../_config/algolia';
import { favoritesFactory } from '../_services/Favorites/Favorites';
import Item from './Item';

const loaderWrapper = {
  position: 'absolute',
  width: '100px',
  height: '50px',
  top: '50%',
  left: '50%',
  'margin-left': '-50px' /* margin is -0.5 * dimension */,
  'margin-top': '-25px',
};

const Loading = (
  <div style={loaderWrapper}>
    <ReactLoading type="bars" color="#F4A261" height="5em" width="5em" />
  </div>
);

const favorites = favoritesFactory();

class ResultsView extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isLoading: false,
      isError: false,
      error: '',
    };
  }

  componentDidMount() {
    // eslint-disable-next-line
    this.setState({ isLoading: true });
    favorites
      .getFavoritesFromAPI()
      .then(() => {
        this.setState({ isLoading: false });
      })
      .catch((err) => {
        this.setState({ isError: true, error: err });
      });
  }

  render() {
    const { result } = this.props;
    const { isLoading, isError, error } = this.state;

    const returnComponent = isLoading ? (
      Loading
    ) : (
      <div className="container-fluid">
        {result.hits.map(item => (
          <Item
            key={item.objectID}
            item={item}
            isFavorite={favorites.isFavorite}
            onAddFavorite={favorites.add}
            onRemoveFavorite={favorites.remove}
          />
        ))}
        <Paginator />
      </div>
    );
    return isError ? <span>{error.toString()}</span> : returnComponent;
  }
}

ResultsView.propTypes = {
  result: PropTypes.shape({
    hits: PropTypes.array,
    page: PropTypes.number,
    nbPages: PropTypes.number,
    searchTerm: PropTypes.string,
  }),
};

ResultsView.defaultProps = {
  result: {
    hits: [],
    page: 0,
    nbPages: 0,
    searchTerm: algolia.DEFAULT_QUERY,
  },
};

export default ResultsView;
