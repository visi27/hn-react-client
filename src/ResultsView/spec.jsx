import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import React from 'react';
import algolia from '../_config/algolia';
import Item from './Item';
import ResultsView from './presenter';

configure({ adapter: new Adapter() });

describe('ResultsView', () => {
  const props = {
    result: {
      hits: [
        {
          title: '1',
          author: '1',
          num_comments: 1,
          points: 2,
          objectID: 'y',
          created_at: '2018-04-03T09:35:47.000Z',
        },
        {
          title: '2',
          author: '2',
          num_comments: 1,
          points: 2,
          objectID: 'z',
          created_at: '2018-04-03T09:35:47.000Z',
        },
      ],
      page: 0,
      nbPages: 0,
      searchTerm: algolia.DEFAULT_QUERY,
    },
  };
  test('has valid snapshot', () => {
    const component = shallow(<ResultsView {...props} />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  it('shows two items on the list', () => {
    const element = shallow(<ResultsView {...props} />);
    expect(element.find(Item).length).toBe(2);
  });
});
