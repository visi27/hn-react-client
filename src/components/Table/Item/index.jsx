import React from 'react';
import { PropTypes } from 'prop-types';
import Button from '../../Button/presenter';
import * as style from './style';

const Item = ({ item, onDismiss }) => (
  <div key={item.objectID} className="container-fluid">
    <span style={style.largeColumn}>
      <a href={item.url}>{item.title}</a>
    </span>
    <div className="list-inline">
      <span className="list-inline-item">Author: {item.author}</span>
      <span className="list-inline-item">Comments: {item.num_comments}</span>
      <span className="list-inline-item">Points: {item.points}</span>
      <span className="list-inline-item">
        <Button className="btn btn-outline-danger btn-sm" onClick={() => onDismiss(item.objectID)}>
          Dismiss
        </Button>
      </span>
    </div>
  </div>
);

Item.propTypes = {
  item: PropTypes.shape({
    created_at: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    title_highlighted: PropTypes.string,
    url: PropTypes.string,
    author: PropTypes.string.isRequired,
    points: PropTypes.number.isRequired,
    story_text: PropTypes.string,
    num_comments: PropTypes.number.isRequired,
    objectID: PropTypes.string.isRequired,
  }).isRequired,
  onDismiss: PropTypes.func,
};

Item.defaultProps = {
  onDismiss: () => {},
};

export default Item;
