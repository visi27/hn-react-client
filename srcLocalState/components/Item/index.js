import React from 'react';
import Button from '../Button/';
import * as style from './style';

const Item = ({ item, onDismiss }) => (
  <div key={item.objectID} className="table-row">
    <span style={style.largeColumn}>
      <a href={item.url}>{item.title}</a>
    </span>
    <span style={style.midColumn}>{item.author}</span>
    <span style={style.smallColumn}>{item.num_comments}</span>
    <span style={style.smallColumn}>{item.points}</span>
    <span style={style.smallColumn}>
      <Button onClick={() => onDismiss(item.objectID)}>Dismiss</Button>
    </span>
  </div>
);

export default Item;
