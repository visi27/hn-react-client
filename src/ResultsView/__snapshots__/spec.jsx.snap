// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsView has valid snapshot 1`] = `
<div
  className="container-fluid"
>
  <Item
    isFavorite={[Function]}
    item={
      Object {
        "author": "1",
        "created_at": "2018-04-03T09:35:47.000Z",
        "num_comments": 1,
        "objectID": "y",
        "points": 2,
        "title": "1",
      }
    }
    key="y"
    onAddFavorite={[Function]}
    onRemoveFavorite={[Function]}
  />
  <Item
    isFavorite={[Function]}
    item={
      Object {
        "author": "2",
        "created_at": "2018-04-03T09:35:47.000Z",
        "num_comments": 1,
        "objectID": "z",
        "points": 2,
        "title": "2",
      }
    }
    key="z"
    onAddFavorite={[Function]}
    onRemoveFavorite={[Function]}
  />
  <Connect(Paginator) />
</div>
`;
